File: 112.Path Sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean dfs(TreeNode root, int currentSum, int target) {
        //check if the tree is empty
        if (root == null) {
            return false;
        }
        //add the sum of the root.val of current node 
        currentSum += root.val;
        //check if the node is leaf node and if the current sum with the target 
        if (root.left == null && root.right == null) {
            return currentSum == target;
        }
        //goes throught left size and goes through the right 
        return dfs(root.left, currentSum, target) ||  dfs(root.right, currentSum, target);
}
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root,0,targetSum);
    }
}

File: 112.Path Sum/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def hasPathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: bool
        """
        def helper(root,curr):
            #empty tree check
            if not root: return False
            curr+=root.val
            #current not is leaf or not check
            if not root.left and not root.right: return targetSum==curr
            return helper(root.left,curr) or helper(root.right,curr)
        return helper(root,0)